### PR TITLE
Feature/alert

### LINKED ISSUE #45: [UI core library] Alert
Originates at https://github.com/hazelcast/frontend-shared/issues/4

https://zeroheight.com/11d0e6dac/p/62d703-alert

Create an Alert component with API:
type
icon
content (as children)

### PR DESCRIPTION
**Pull request by jankoritak**
Closes #45 

TODO:
- [x] Button action type
- [x] Weird actions alignment

### CODE REVIEW

File: packages/ui/__stories__/Alert.stories.tsx
Comment (aigoncharov) [inline packages/ui/__stories__/Alert.stories.tsx:1]: Great set of stories!

File: packages/ui/__stories__/Icon.stories.tsx
Comment (aigoncharov) [inline packages/ui/__stories__/Icon.stories.tsx:1]: Thank you for doing this!

File: packages/ui/src/Alert.tsx
Comment (aigoncharov) [inline packages/ui/src/Alert.tsx:34]: This is what I call well-thought types! Good job!

File: packages/ui/src/Toast.module.scss
Comment (aigoncharov) [inline packages/ui/src/Toast.module.scss:17]: Is there any difference?

File: packages/ui/src/Button.module.scss
Comment (aigoncharov) [inline packages/ui/src/Button.module.scss:127]: Seems like we do not need `color: c.$colorNeutralWhite;` for the `.button` class. If we removed it we could use here `color: inherit` and let the parent container define the color of the button text. It would allow this button to have proper visuals in any parent context. WDYT?

File: packages/ui/src/Button.tsx
Comment (aigoncharov) [inline packages/ui/src/Button.tsx:18]: nit: Do you think we have a use-case for these extra props right now? Our buttons at this point are of fixed height and colors. So the icons are also of fixed height and colors.

File: packages/ui/src/Button.tsx
Comment (aigoncharov) [inline packages/ui/src/Button.tsx:117]: Good thinking on using the `Icon` component!

File: packages/ui/src/Alert.module.scss
Comment (aigoncharov) [inline packages/ui/src/Alert.module.scss:50]: If we went along with https://github.com/hazelcast/frontend-shared/pull/49/files#r511764453 we could probably remove this class at all

File: packages/ui/src/Alert.module.scss
Comment (aigoncharov) [inline packages/ui/src/Alert.module.scss:29]: The parent container here is flex. Why do we need to make this element absolute? Wouldn't flex parent position it correctly if we added `flex: 1 0 auto;` to `title`?

File: packages/ui/src/Alert.tsx
Comment (aigoncharov) [inline packages/ui/src/Alert.tsx:65]: I am wondering if we should add `transparent` type here as well. We also need to think if `#FAFAFA` background works as the hover effect. @pawelklasa WDYT?
The current design of Icon Button (https://www.figma.com/file/8mVm6LTbp2Z0RaWWjTZoft/%F0%9F%90%9DHIVE---Hazelcast-Design-System?node-id=3352%3A8) does not play nice with different background colors.
We would like to use it for the `X` button for our alerts https://www.figma.com/file/8mVm6LTbp2Z0RaWWjTZoft/%F0%9F%90%9DHIVE---Hazelcast-Design-System?node-id=2341%3A111, but the `#FAFAFA` hover does not fit there. Should we do something else for the hover effect? Maybe, we could change the color instead? E.g. for the hover. of the `neutral-white` Icon Button we could make it `neutral0lighter`, for the `colot-text` we could do `secondary-text`. Alternatively, we could darker the background. WDYT?

File: packages/ui/src/Button.module.scss
Comment (jankoritak) [reply packages/ui/src/Button.module.scss:126]: Sure, we can do that!

File: packages/ui/src/Toast.module.scss
Comment (jankoritak) [reply packages/ui/src/Toast.module.scss:17]: None. I honestly don't remember the reasoning behind this. Can roll back, if you prefer to include it in `.content`.

File: packages/ui/src/Button.tsx
Comment (jankoritak) [reply packages/ui/src/Button.tsx:18]: These are defined because we need to tweak the spacing a bit for Alerts actions: https://github.com/hazelcast/frontend-shared/pull/49/files#diff-674337aa6f682273f1632c6b0625a49e0606aa8f8e06facd4ca62a475ed76231R86.

File: packages/ui/src/Alert.module.scss
Comment (jankoritak) [reply packages/ui/src/Alert.module.scss:28]: Good thinking!

File: packages/ui/src/Alert.module.scss
Comment (jankoritak) [reply packages/ui/src/Alert.module.scss:49]: Can't get the link working properly. Could you post a snippet of what do you mean?

File: packages/ui/src/Alert.tsx
Comment (jankoritak) [reply packages/ui/src/Alert.tsx:65]: I would lean towards changing the `IconButton` hover/rollover. If we can find an accessible alternative, to this `background-color` toggle, great! If not, we can surely add `transparent` `IconButton` type. This is makes sense to me, as we already implement it for regular button, right?

I'm a. bit lost in the last couple of sentences, though.

> Maybe, we could change the color instead? E.g. for the hover. of the neutral-white Icon Button we could make it neutral0lighter, for the colot-text we could do secondary-text. Alternatively, we could darker the background. WDYT?

`IconButton` does not have text, right?

Comment (aigoncharov) [reply packages/ui/src/Alert.tsx:65]: @jankoritak I meant that for the hover effect we could simply darken the background or darken the icon color.

File: packages/ui/src/Toast.module.scss
Comment (aigoncharov) [reply packages/ui/src/Toast.module.scss:17]: I asked just out of curiosity. No need to rollback :)

File: packages/ui/src/Alert.module.scss
Comment (aigoncharov) [reply packages/ui/src/Alert.module.scss:46]: It was linking to the idea of using `color: inherit` for the transparent button.

File: packages/ui/src/Button.tsx
Comment (aigoncharov) [reply packages/ui/src/Button.tsx:18]: Oh, I see. We not only decreased the margin there, but also decreased the font size of the button. Do you think it makes sense to add a `size` prop to the `Button` with possible values `'small' | 'normal'`. For the `small` `Button` we could decrease the margins, change the icon sizes and remove capitalizations. In other words, kind of encapsulate these styling changes. WDYT?

Comment (aigoncharov) [reply packages/ui/src/Button.tsx:18]: I think I am basically building on the previous idea we discussed during the initial `Button` implementation. Instead of having lots of props with lots of freedom, we should try created several more or less rigid presets. WDYT?

File: packages/ui/styles/constants/_typography.scss
Comment (aigoncharov) [inline packages/ui/styles/constants/_typography.scss:74]: I would suggest to parameterize  `typographyBodyNormal`
```scss
@mixin typographyBodyNormal($weight: 'regular') {
  font-family: $fontFamilyText;
  font-size: $fontSizeBodyNormal;
  letter-spacing: $letterSpacingBodyNormal;
  line-height: $lineHeightBodyNormal;

  @if $weight == 'regular' {
    font-weight: $fontWeightTextRegular; 
  }
  @if $weight == 'bold' {
    font-weight: $fontWeightTextBold; 
  }
  @if $weight == 'light' {
    font-weight: $fontWeightTextLight; 
  }
}	}
```

Comment (aigoncharov) [reply packages/ui/styles/constants/_typography.scss:74]: In fact, we could go even further
```scss
@mixin typographyBodyWeight($weight: 'regular') {
  @if $weight == 'regular' {
    font-weight: $fontWeightTextRegular; 
  }
  @if $weight == 'bold' {
    font-weight: $fontWeightTextBold; 
  }
  @if $weight == 'light' {
    font-weight: $fontWeightTextLight; 
  }
}

@mixin typographyBodyNormal($weight: 'regular') {
  font-family: $fontFamilyText;
  font-size: $fontSizeBodyNormal;
  letter-spacing: $letterSpacingBodyNormal;
  line-height: $lineHeightBodyNormal;

  @include typographyBodyWeight($weight);
}	}
```

File: packages/ui/styles/constants/_typography.scss
Comment (aigoncharov) [inline packages/ui/styles/constants/_typography.scss:45]:
```suggestion
```

File: packages/ui/src/Alert.tsx
Comment (jankoritak) [reply packages/ui/src/Alert.tsx:65]: I see now. Let's wait for @pawelklasa's POV.

File: packages/ui/styles/constants/_typography.scss
Comment (jankoritak) [reply packages/ui/styles/constants/_typography.scss:74]: Great idea!

File: packages/ui/src/Button.tsx
Comment (jankoritak) [reply packages/ui/src/Button.tsx:18]: Conceptually, that makes sense to me, however are we sure that we'll ever use `small` button in any case expect from this one? I would defer this if we have another use-case.

Comment (aigoncharov) [reply packages/ui/src/Button.tsx:18]: Fair enough. Let's not add potentially semi-dead code.

**Review by aigoncharov — APPROVED** [review]
LGTM!
I think we should not wait for the transparent IconButton design from Pawel and merge this one in. Once we have the design we will refactor.
I would suggest to address https://github.com/hazelcast/frontend-shared/pull/49/files#diff-4a2ce829e49e0bb7a1b6dfdb6fb013be8ed6728646e50bd2707e5e43fd05fab0R46 before merging.
Could you also generate the reference screenshots for the news stories?

File: packages/ui/src/Alert.module.scss
Comment (jankoritak) [reply packages/ui/src/Alert.module.scss]: Was able to get rid of the class in the end. Addressed in: https://github.com/hazelcast/frontend-shared/pull/49/commits/87c15d54aaccef3c99a0047b1eae2ec54c719544

**Pull request merged**
